File: java/doit_java/Chapter8/src/witharraylist/Customer.java
// 고객 관리 프로그램 완성하기 1
package witharraylist;

// inheritance Customer5를 복사해옴
public class Customer {
    protected int customerId;       // 고객 아이디
    protected String customerName;  // 고객 이름
    protected String customerGrade; // 고객 등급
    int bonusPoint;                 // 보너스 포인트
    double bonusRatio;              // 적립 비율

    // 새로운 디폴트 생성자
    public Customer() {
        initCustomer();     // 고객 등급과 보너스 포인트 적립률 지정 함수 호출
    }

    public Customer(int customerId, String customerName) {
        this.customerId = customerId;
        this.customerName = customerName;
        // customerGrade = "SILVER";
        // bonusRatio = 0.01;
        initCustomer();     // 고객 등급과 보너스 포인트 적립률 지정 함수 호출
    }
    /* 원래의 디폴트 생성자
    public Customer() {
        customerGrade = "SILVER";
        bonusRatio = 0.01;
    }*/
    // 디폴트 생성자의 내용을 private로 변경하고 멤버변수 초기화 부분을 호출하도록 변경
    private void initCustomer() {
        customerGrade = "SILVER";   // 기본 등급
        bonusRatio = 0.01;          // 보너스 포인트 기본 적립 비율
    }
    // 보너스 포인트 적립, 지불 가격 계산 메서드
    public int calcPrice(int price) {
        bonusPoint += price * bonusRatio;   // 보너스 포인트 계산
        return price;
    }
    // 고객 정보를 반환하는 메서드
    public String showCustomerInfo() {
        return customerName + " 님의 등급은 " + customerGrade + " 이며, 보너스 포인트는 " + bonusPoint + " 입니다.";
    }
    // protected 예약어로 선언한 변수를 외부에서 사용할수 있도록 get(), set() 메서드 추가
    public int getCustomerId() {
        return customerId;
    }
    public void setCustomerId(int customerId) {
        this.customerId = customerId;
    }
    public String getCustomerName() {
        return customerName;
    }
    public void setCustomerName(String customerName) {
        this.customerName = customerName;
    }
    public String getCustomerGrade() {
        return customerGrade;
    }
    public void setCustomerGrade(String customerGrade) {
        this.customerGrade = customerGrade;
    }
}

File: python/p_oop2/chap07/inheritance/customer.py
class Customer:
    def __init__(self):
        self.customer_id = 0;
        self.customer_name = "";
        self.customer_grade = "SILVER"
        self.bonus_point = 0
        self.bonus_ratio = 0.01

    def calc_price(self, price):
        self.bonus_point += price * self.bonus_point
        return price

    def show_customer_info(self):
        return self.customer_name + "님의 등급은 " + self.customer_grade + "이며, 적립된 보너스 포인트는 " + str(self.bonus_point) + "점 입니다."

    def get_customer_id(self):
        return customer_id

    def set_customer_id(self, customer_id):
        self.customer_id = customer_id

    def get_customer_name(self):
        return customer_name

    def set_customer_name(self, customer_name):
        self.customer_name = customer_name

    def get_customer_grade(self):
        return customer_grade

    def set_customer_grade(self, customer_grade):
        this.customer_grade = customer_grade
